### PR TITLE
[Enhancement] handle const column in array_filter

### PR DESCRIPTION
**Pull request by silverbullet233**
## What I'm doing:

The array_filter function needs to unpack all const columns before evaluation. In fact, we can do some special handling on them.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exprs/array_functions.tpp [hunk 1]
```diff
@@ -904,65 +906,107 @@ private:
...
+    static ColumnPtr _array_filter(const ColumnPtr& src_column, const ColumnPtr& filter_column) {
+        if (src_column->only_null()) {
+            return src_column;
+        }
+        bool is_src_const = src_column->is_constant();
+        bool is_filter_const = filter_column->is_constant();
+
+        size_t chunk_size = src_column->size();
+        if (filter_column->only_null()) {
+            if (is_src_const) {
```
Comment (trueeyu) [inline be/src/exprs/array_functions.tpp:924]: Why not process const column in prepare?

File: be/src/exprs/array_functions.tpp [hunk 2]
```diff
@@ -904,65 +906,107 @@ private:
...
+                    const auto src_null_column = down_cast<const NullableColumn*>(src_column.get())->null_column();
+                    auto dest_nullable_column = down_cast<NullableColumn*>(dest_column.get());
+                    auto dest_null_column = dest_nullable_column->mutable_null_column();
+                    dest_null_column->get_data().assign(src_null_column->get_data().begin(),
+                                                        src_null_column->get_data().end());
+                    dest_nullable_column->set_has_null(src_column->has_null());
+                    data_column = dest_nullable_column->data_column();
+                }
+                data_column->append_default(chunk_size);
+                return dest_column;
```
Comment (trueeyu) [inline be/src/exprs/array_functions.tpp:944]: Why not use data_column?

File: be/src/exprs/array_functions.tpp [hunk 3]
```diff
@@ -904,65 +906,107 @@ private:
...
+            return dest_column;
         }
-        return dest_column;
     }
 
+    template <bool ConstSrc, bool ConstFilter>
     static void _filter_array_items(const ArrayColumn* src_column, const ColumnPtr& raw_filter,
-                                    ArrayColumn* dest_column, NullColumn* dest_null_map) {
+                                    ArrayColumn* dest_column, NullColumn* dest_null_map, size_t src_rows) {
+        if constexpr (!ConstSrc) {
```
Comment (trueeyu) [inline be/src/exprs/array_functions.tpp:1007]: No need to add `if constexpr (!ConstSrc)`?

**Review by trueeyu — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by silverbullet233** [comment]
@mergify backport branch-3.2

**Pull request merged**